contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -742,51 +768,176 @@ have a built-in `nil` value.
 
 ## Defaulting
 
-Default resource values are API version-specific, and they are applied during
-the conversion from API-versioned declarative configuration to internal objects
-representing the desired state (`Spec`) of the resource. Subsequent GETs of the
-resource will include the default values explicitly.
-
-Incorporating the default values into the `Spec` ensures that `Spec` depicts the
-full desired state so that it is easier for the system to determine how to
-achieve the state, and for the user to know what to anticipate.
+In general we want default values to be explicitly represented in our APIs,
+rather than asserting that "unspecified fields get the default behavior".  This
+is important so that:
+ - default values can evolve and change in newer API versions
+ - the stored configuration depicts the full desired state, making it easier
+   for the system to determine how to achieve the state, and for the user to
+   know what to anticipate
+
+There are 3 distinct ways that default values can be applied when creating or
+updating (including patch and apply) a resource:
+
+ 1. static: based on the requested API version and possibly other fields in the
+    resource, fields can be assigned values during the API call
+ 2. admission control: based on the configured admission controllers and
+    possibly other state in or out of the cluster, fields can be assigned
+    values during the API call
+ 3. controllers: arbitrary changes (within the bounds of what is allowed) can
+    be made to a resource after the API call has completed
+
+Some care is required when deciding which mechanism to use and managing the
+semantics.
```

Comment:
+1 to this statement. I am concerned about developers making the right selection here.

Can we offer any more guidance on when to chose which option?  I feel like using static when possible is obvious enough, but maybe the trade-offs between (2) and (3) can be articulated a bit more in this doc to help developer make sound choices?  (my personal bias is to encourage developers to avoid admission control unless it's really essential, so I'd be curious what use cases really demand it and why)

Reply:
The drawback of the "asynchronous default", i..e option (3), is that it temporarily leaves the object in a state where the value is indeterminate and all other controllers must be written with that in mind.

Reply:
I agree.  This is where I struggle with semantics.  If it's something controller sets, I find the term "default" a bit odd.

A lot of controllers should be designed to wait until resources are in the state where they can do their task. So what one controller might consider "initialized", might be mid-way or even pretty late in some resources lifecycle, and might be very different from what another controller that actuates on the same resource considers "initialized".

Reply:
Yeah, (3) isn't really default - I'll clarify that.